extensions/sql/asset/index/src/main/java/org/eclipse/dataspaceconnector/sql/asset/index/SqlAssetIndex.java

```diff
@@ -127,23 +125,53 @@ public Asset deleteById(String assetId) {
         }
     }
 
+    @Override
+    public void accept(AssetEntry item) {
+        Objects.requireNonNull(item);
+
+        accept(item.getAsset(), item.getDataAddress());
+    }
+
     @Override
     public Stream<Asset> queryAssets(AssetSelectorExpression expression) {
         Objects.requireNonNull(expression);
 
-        return getAssetStream(sqlAssetQueries.getSqlAssetListClause());
+        var criteria = expression.getCriteria();
+        var querySpec = QuerySpec.Builder.newInstance().filter(criteria)
+                .offset(0)
+                .limit(Integer.MAX_VALUE) // means effectively no limit
+                .build();
+        return queryAssets(querySpec);
     }
 
     @Override
     public Stream<Asset> queryAssets(QuerySpec querySpec) {
         Objects.requireNonNull(querySpec);
 
-        var query = String.format("%s LIMIT %s OFFSET %s",
-                sqlAssetQueries.getSqlAssetListClause(),
-                querySpec.getLimit(),
-                querySpec.getOffset());
 
-        return getAssetStream(query);
+        var conditions = querySpec.getFilterExpression().stream().map(SqlConditionExpression::new).collect(Collectors.toList());
```

Comment:
Is it possible that this piece of code can be reused in different stores? 
Maybe it would be worth to extract it to separate class (called eg. SQLQueryResolver) that would implement QueryResolver interface. See eg. ReflectionBasedQueryResolver.
_[I see the PR was just merged but will leave the comment for future reference]_

Reply:
again, this is really specific to the `AssetIndex`, where all the properties are located in a separate table. However, I'm sure we can extract some useful code when we need it, especially the `SqlConditionExpression`, which "parses" the `Criterion`